### PR TITLE
Feat : Get indexed items total count from Algolia

### LINKED ISSUE #185: Get indexed items total count from Algolia
**Is your feature request related to a problem? Please describe.**
For consistency we want to get the total count of indexed items from the source -- Algolia.

**Describe the solution you'd like**
Create a one method to determine the total count of items for all entities (issue, project, chapter, committee). Cache the value for better performance. Name the field `total_items`. 

**Additional context**
Clean up the currently used method if they are no longer used.

### PR DESCRIPTION
**Pull request by Rajgupta36**
Resolves #185
---
## Overview

This pull request aims to replace the existing index count logic with the use of the Algolia search client. The updates include a new method to get the count of records in the Algolia index and refactoring the models and search endpoint to accommodate the new logic.

---

## Checklist
- [x] Replaced old method with a new method using the Algolia search client.
- [x] Implemented `get_algolia_index_count` method.
- [x] Updated the search endpoint with the new methods.
- [x] Tested the new search functionality thoroughly.

### CODE REVIEW

File: backend/apps/common/index.py [hunk 1]
```diff
@@ -14,12 +15,19 @@ class IndexSynonymsMixin:
     """Nest index synonyms mixin."""
 
     @staticmethod
-    def reindex_synonyms(app_name, index_name):
-        """Reindex synonyms."""
-        index = SearchClient.create(
+    def _get_algolia_client():
```
Comment (arkid15r) [inline backend/apps/common/index.py]:
```suggestion
    def _get_client():
```

File: backend/apps/common/index.py [hunk 2]
```diff
@@ -38,3 +46,15 @@ def reindex_synonyms(app_name, index_name):
 
         index.clear_synonyms()
         index.save_synonyms(synonyms, {"replaceExistingSynonyms": True})
+
+    @staticmethod
+    @lru_cache(maxsize=1024)
+    def get_algolia_index_count(index_name):
```
Comment (arkid15r) [inline backend/apps/common/index.py]:
```suggestion
    def get_total_count(index_name):
```

File: backend/apps/owasp/api/search/chapter.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 from django.http import JsonResponse
 
 from apps.common.geocoding import get_ip_coordinates
+from apps.common.index import IndexSynonymsMixin
```
Comment (arkid15r) [inline backend/apps/owasp/api/search/chapter.py]: This needs to be renamed into `IndexBase` or similar as we extended it beyond synonyms.

File: backend/apps/owasp/api/search/chapter.py [hunk 4]
```diff
@@ -39,11 +40,11 @@ def chapters(request):
     """Search chapters API endpoint."""
     page = int(request.GET.get("page", 1))
     query = request.GET.get("q", "")
-
+    active_chapters_count = IndexSynonymsMixin.get_algolia_index_count("local_chapters")
```
Comment (arkid15r) [inline backend/apps/owasp/api/search/chapter.py]: The `local_*` naming will only work for local environment. 
Please update the base class to add environment prefix to the index name and pass only `chapters` (same for other indexed models).

File: backend/apps/owasp/api/search/chapter.py [hunk 5]
```diff
@@ -39,11 +40,11 @@ def chapters(request):
...
     chapters = get_chapters(query=query, page=page, meta=request.META)
     return JsonResponse(
         {
-            "active_chapters_count": Chapter.active_chapters_count(),
+            "active_chapters_count": active_chapters_count,
```
Comment (arkid15r) [inline backend/apps/owasp/api/search/chapter.py]: You can inline the value, like:
```suggestion
            "active_chapters_count": IndexBase.get_total_count("chapters")
```

**Review by arkid15r — CHANGES_REQUESTED** [review]
Here is some initial feedback:

File: backend/apps/common/index.py [hunk 6]
```diff
@@ -38,3 +44,17 @@ def reindex_synonyms(app_name, index_name):
...
+    @lru_cache(maxsize=1024)
+    def get_total_count(index_name):
+        """Get  total count of records in an Algolia index."""
+        try:
+            index = IndexBase._get_client().init_index(
+                f"{settings.ENVIRONMENT.lower()}_{index_name}"
+            )
+            return index.search("", {"hitsPerPage": 0, "analytics": False})["nbHits"]
+
+        except Exception:
```
Comment (arkid15r) [inline backend/apps/common/index.py]: Avoid using too broad exceptions. If you don't need it now just remove it.

File: backend/apps/github/models/issue.py [hunk 7]
```diff
@@ -180,12 +178,6 @@ def bulk_save(issues, fields=None):
         """Bulk save issues."""
         BulkSaveModel.bulk_save(Issue, issues, fields=fields)
 
-    @staticmethod
-    @lru_cache
-    def open_issues_count():
-        """Return open issues count."""
-        return Issue.open_issues.assignable.count()
-
```
Comment (arkid15r) [inline backend/apps/github/models/issue.py]: As you removed existing method you run tests to check it still works.
The problem is that Nest isn't fully covered with tests. NestBot still uses these methods and need to be updated too. So here you have 2 options:
  - keep the old methods and their signatures while updating their logic with `IndexBase` stuff
  - replace old methods with `IndexBase.get_total_count` everywhere by using search

**Review by arkid15r — COMMENTED** [review]
Looks better. We also should address:

File: backend/apps/common/index.py (hunk 6 above)
Comment (Rajgupta36) [reply backend/apps/common/index.py]: okkk

File: backend/apps/owasp/models/chapter.py [hunk 8]
```diff
@@ -61,12 +59,6 @@ def __str__(self):
         """Chapter human readable representation."""
         return f"{self.name or self.key}"
 
-    @staticmethod
-    @lru_cache
-    def active_chapters_count():
-        """Return active chapters count."""
-        return Chapter.active_chapters.count()
```
Comment (arkid15r) [inline backend/apps/owasp/models/chapter.py:68]: The issues refactoring approach needs to be kept with other models too.

**Review by arkid15r — CHANGES_REQUESTED** [review]
You need to refactor all existing methods to the new IndexBase approach. Keep the methods, change the logic.

File: backend/apps/owasp/models/chapter.py (hunk 8 above)
Comment (Rajgupta36) [reply backend/apps/owasp/models/chapter.py:68]: ohh i didn't notice sorry

File: backend/apps/owasp/api/search/chapter.py [hunk 9]
```diff
@@ -39,11 +40,10 @@ def chapters(request):
     """Search chapters API endpoint."""
     page = int(request.GET.get("page", 1))
     query = request.GET.get("q", "")
-
     chapters = get_chapters(query=query, page=page, meta=request.META)
     return JsonResponse(
         {
-            "active_chapters_count": Chapter.active_chapters_count(),
+            "active_chapters_count": IndexBase.get_total_count("chapters"),
```
Comment (arkid15r) [inline backend/apps/owasp/api/search/chapter.py]: It seems you can revert these too.

Comment (Rajgupta36) [reply backend/apps/owasp/api/search/chapter.py]: yes

File: backend/apps/common/index.py [hunk 10]
```diff
@@ -1,25 +1,32 @@
...
-    """Nest index synonyms mixin."""
+class IndexBase:
+    """Nest index synonyms mixin and record count."""
 
     @staticmethod
-    def reindex_synonyms(app_name, index_name):
-        """Reindex synonyms."""
-        index = SearchClient.create(
+    def _get_client():
+        """Reusable method to get the Algolia client."""
```
Comment (arkid15r) [inline backend/apps/common/index.py]:
```suggestion
        """Get the Algolia client."""
```

File: backend/apps/common/index.py [hunk 11]
```diff
@@ -38,3 +45,17 @@ def reindex_synonyms(app_name, index_name):
 
         index.clear_synonyms()
         index.save_synonyms(synonyms, {"replaceExistingSynonyms": True})
+
+    @staticmethod
+    @lru_cache(maxsize=1024)
+    def get_total_count(index_name):
+        """Get  total count of records in an Algolia index."""
```
Comment (arkid15r) [inline backend/apps/common/index.py]:
```suggestion
        """Get total count of records in index."""
```

**Review by arkid15r — APPROVED** [review]
LGTM 👍

**Pull request merged**